Refactor client lockout parameter resolution

hirotasoshu opened the pull request:
Resolves #692 and #690

hirotasoshu commented on axes/utils.py:
```diff
@@ -37,23 +37,38 @@ def reset_request(request: HttpRequest) -> int:
     username = request.GET.get("username", None)
 
-    ip_or_username = settings.AXES_LOCK_OUT_BY_USER_OR_IP
-    if settings.AXES_ONLY_USER_FAILURES:
+    ip_required = False
+    username_required = False
+    ip_and_username = False
+
+    for param in lockout_paramaters:
+        # hack: in works with all iterables, including strings
+        # so this checks works with separate parameters
+        # and with parameters combinations
+        if "username" in param and "ip_address" in param:
+            ip_and_username = True
+            ip_required = True
+            username_required = True
+            break
+        elif "username" in param:
+            username_required = True
+        elif "ip_address" in param:
+            ip_required = True
+
+    ip_or_username = not ip_and_username and ip_required and username_required
+    if not ip_required:
         ip = None
-    elif not (
-        settings.AXES_LOCK_OUT_BY_USER_OR_IP
-        or settings.AXES_LOCK_OUT_BY_COMBINATION_USER_AND_IP
-    ):
+    if not username_required:
```
That's kindy ugly, I think we should change reset API

hirotasoshu commented on axes/apps.py:
```diff
@@ -33,14 +33,22 @@ def initialize(cls):
+
             else:
-                mode = "blocking by IP only"
+                mode = "blocking by " + " or ".join(
+                    [
+                        param
+                        if isinstance(param, str)
+                        else "combination of " + " and ".join(param)
+                        for param in settings.AXES_LOCKOUT_PARAMETERS
+                    ]
+                )
+
```
This produces almost the same log messages as we had before (the only difference that it IP word changed to ip_address and messages does not have word only if we have only one lockout parameter)

hirotasoshu commented on axes/handlers/cache.py:
```diff
@@ -78,9 +79,10 @@ def user_login_failed(self, sender, credentials: dict, request=None, **kwargs):
 
         username = get_client_username(request, credentials)
-        if settings.AXES_ONLY_USER_FAILURES and username is None:
+        lockout_parameters = list(get_lockout_parameters(request, credentials))
```
Since AXES_LOCKOUT_PARAMETERS could be any iterable, for simplicity, I convert it to list

hirotasoshu commented on axes/helpers.py:
```diff
@@ -226,26 +255,35 @@ def get_client_parameters(username: str, ip_address: str, user_agent: str) -> li
-            filter_query[0]["user_agent"] = user_agent
+    filter_kwargs = []
+
+    for parameter in lockout_parameters:
+        try:
+            if isinstance(parameter, str):
+                filter_kwarg = {parameter: parameters_dict[parameter]}
+            else:
+                filter_kwarg = {
+                    combined_parameter: parameters_dict[combined_parameter]
+                    for combined_parameter in parameter
+                }
+            filter_kwargs.append(filter_kwarg)
+
+        except KeyError as e:
+            error_msg = (
+                f"{e} lockout parameter is not allowed. "
+                f"Allowed parameters: {', '.join(parameters_dict.keys())}"
+            )
+            raise ValueError(error_msg) from e
```
Should we raise error like this or just skip incorrect parameters?
Or maybe we can check parameters for correctness in `get_lockout_parameters`?

hirotasoshu commented on tests/test_attempts.py:
```diff
@@ -82,74 +82,74 @@ def test_reset_ip_username(self):
 
-    @override_settings(AXES_LOCK_OUT_BY_USER_OR_AND=True)
+    @override_settings(AXES_LOCKOUT_PARAMETERS=[("username", "ip_address")])
     def test_reset_ip_username_user_and_ip(self):
         self.request.GET["username"] = self.USERNAME_1
         self.request.META["REMOTE_ADDR"] = self.IP_1
         reset_request(self.request)
-        self.assertEqual(AccessAttempt.objects.count(), 3)
+        self.assertEqual(AccessAttempt.objects.count(), 4)
```
This test was incorrect

hirotasoshu commented:
Tests broke in old python versions, because `Iterable` before python3.8 was in collections module, not collections.abc

hirotasoshu commented:
@aleksihakli I left a couple comments for this implementation. If everything is seems ok to you, I will add new tests (since now lockout parameters and we need additional tests to ensure everything is working fine) and update documentation

aleksihakli commented:
Thanks for the PR! The approach looks good and backwards compatible which is nice. I'll try and wrap up a review soon.

aleksihakli commented:
Looks good to me.

I'd maybe adjust a few small things either before or after merging, it's easy to make separete update commits or PRs on top of this after upstreaming too:

- change the types for the: use nested lists of instead of tuples inside lists, just a personal preference for the sake of simplicity
- add a `log.exception` message on invalid lockout keys / values for sake of debuggability
- add a test for triggering the `KeyError` / `ValueError` for sake of verification (maybe test the log trace as well in the same place)
- add a deprecation check for anyone that might use the new and old flags in a mixed fashion by accident, and
- add a few more docstring examples for the new lockout param introductions and available variants

but these are just small nuances and preferences.

I think this PR implements the setting and API updates in a more flexible and more understandable fashion than the current stack does as it currently is. It's a good pragmatic approach for modernizing the implementation 👍

Thanks for the effort. Let's get this merged and update instructions written so we can roll out 6.0 soon 💪

aleksihakli commented:
Current `prospector` errors to-fix:

https://github.com/jazzband/django-axes/actions/runs/4881413414/jobs/8710287094?pr=1050#step:7:91

Current typing errors to-fix:

https://github.com/jazzband/django-axes/actions/runs/4881413414/jobs/8710285656?pr=1050#step:7:154

hirotasoshu commented:
@aleksihakli I adjusted types for ``AXES_LOCKOUT_PARAMETERS``: it's now list of strings or nested lists of strings, added log.exception for cases when incorrect lockout parameters passed and also added new tests

aleksihakli commented:
Thank you for the effort, this is a great update to get into the 6.x release to refresh the API and settings and it even has backwards compatibility which is awesome👍I'll go ahead and merge this and bake a new beta release tag, let's see if we have anything to add after that before rolling out 6.0.0 🎉
